### PR TITLE
CART-89 mercury: Update to mercury_2.0.0

### PR DESCRIPTION
**Pull request by frostedcmos**
- Update to mercury v2.0.0

Signed-off-by: Alexander Oganezov <alexander.a.oganezov@intel.com>

### CODE REVIEW

File: utils/rpms/daos.spec [hunk 1]
```diff
@@ -4,9 +4,9 @@
 
 %if (0%{?suse_version} >= 1500)
 # until we get an updated mercury build on 15.2
-%global mercury_version 2.0.0~rc1-1.suse.lp151
+%global mercury_version 2.0.0~rc2-1.suse.lp152
 %else
-%global mercury_version 2.0.0~rc1-1%{?dist}
+%global mercury_version 2.0.0~rc2-1%{?dist}
 %endif
```
Comment (brianjmurrell) [inline utils/rpms/daos.spec:10]: We should have this by now and can get rid of the leap branch.  Hopefully before we ever need it again, we get on a more normal _mercury_ consumer train.

```suggestion
%global mercury_version 2.0.0~rc2-1%{?dist}
```

**Review by brianjmurrell — DISMISSED** [review]
Functional as is, but would be nice to get rid of the gate, and the (`lp15?`) tripping hazard.

**Review by jolivier23 — APPROVED** [review]

File: utils/rpms/daos.spec [hunk 2]
```diff
@@ -2,16 +2,11 @@
...
-# until we get an updated mercury build on 15.2
-%global mercury_version 2.0.0~rc2-1.suse.lp152
-%else
 %global mercury_version 2.0.0~rc2-1%{?dist}
-%endif
 
 Name:          daos
 Version:       1.1.1
-Release:       3%{?relval}%{?dist}
+Release:       4%{?relval}%{?dist}
```
Comment (brianjmurrell) [inline utils/rpms/daos.spec:9]: I think Release: 4 has already landed, unfortunately.

**Review by brianjmurrell — APPROVED** [review]
Looks good.  Maybe I am wrong about Release: 4 having already landed.

**Comment by liuxuezhao** [comment]
@frostedcmos  mercury v2.0.0rc2 with some changes with OFI RMA transferring, the new code seems will use multi-segments/iovs RMA transferring, that possibly cause PSM2 bulk transferring perf downgrade.
Did someone test if the multi-segments bulk transferring over PSM2 changed on the new version? the multi-sgement bulks means for example client create bulk handle by sgl with 4 iov segments, and server create bulk handle by sgl with 2 or 4 iov segments, to see if bulk bandwidth changed for current mercury version and new version.

**Comment by frostedcmos** [comment]
> Did someone test if the multi-segments bulk transferring over PSM2 changed on the new version? the multi-sgement bulks means for example client create bulk handle by sgl with 4 iov segments, and server create bulk handle by sgl with 2 or 4 iov segments, to see if bulk bandwidth changed for current mercury version and new version.

We have not done perf testing of this rc2. One of expected changing is multi-segment bulk transfer; at this point we are trying to get all functional tests to pass. We should indeed test performance difference once we resolve all functional issues.

**Comment by liuxuezhao** [comment]
> > Did someone test if the multi-segments bulk transferring over PSM2 changed on the new version? the multi-sgement bulks means for example client create bulk handle by sgl with 4 iov segments, and server create bulk handle by sgl with 2 or 4 iov segments, to see if bulk bandwidth changed for current mercury version and new version.
> 
> We have not done perf testing of this rc2. One of expected changing is multi-segment bulk transfer; at this point we are trying to get all functional tests to pass. We should indeed test performance difference once we resolve all functional issues.

OK, it possibly only affect multi-segments bulk performance, that is essential for EC obj IO perf.

File: debian/control [hunk 3]
```diff
@@ -10,7 +10,7 @@ Build-Depends: debhelper (>= 10),
                libopenmpi-dev,
                libssl-dev,
                libyaml-dev,
-               libmercury-dev (= 2.0.0~rc1-1),
+               libmercury-dev (= 2.0.0~rc3-1),
```
Comment (brianjmurrell) [inline debian/control:13]: Could use >= here, per the general review comment:
```suggestion
               libmercury-dev (>= 2.0.0~rc3-1),
```

**Review by brianjmurrell — APPROVED** [review]
Using the `>=` options for the RPM and DEB build would reduce the number of places that need updating (to just `build.config`) when you want to do a version update.  It would also make post-1.2 updating of mercury possible without having to provide new daos* packages also.

File: debian/control (hunk 3 above)
Comment (frostedcmos) [reply debian/control]: done

File: debian/control [hunk 4]
```diff
@@ -10,7 +10,7 @@ Build-Depends: debhelper (>= 10),
                libopenmpi-dev,
                libssl-dev,
                libyaml-dev,
-               libmercury-dev (= 2.0.0~rc3-1),
+               libmercury-dev (>= 2.0.0~rc3-1),
```
Comment (brianjmurrell) [inline debian/control:13]: You'd probably want to do this in daos.spec also.

**Review by brianjmurrell — APPROVED** [review]

File: debian/control (hunk 3 above)
Comment (jolivier23) [inline debian/control:13]: I think there is a debian log that needs updating

**Review by jolivier23 — CHANGES_REQUESTED** [review]

**Comment by frostedcmos** [comment]
1 Failure is due to DAOS-6123 "osa online reintegration timed out"

**Review by brianjmurrell — APPROVED** [review]
Just a head's up that this is going to conflict with 1.1.2-1 versioning update that is in CI right now.

Note for gatekeeper: The version being updated to is not what is in the description (and probably first commit message).

**Review by jolivier23 — DISMISSED** [review]
The title apparently needs to change now that it's not rc2

**Comment by frostedcmos** [comment]
> The title apparently needs to change now that it's not rc2

title and message updated

**Review by brianjmurrell — APPROVED** [review]

**Comment by johannlombardi** [comment]
This needs to be refreshed, i am afraid.

**Comment by johannlombardi** [comment]
I asked @Nasf-Fan to look at the failure and it sounds like this mercury upgrade is adding a regression.

**Comment by Nasf-Fan** [comment]
We hit the following corruption on server side when client transfers small sized data (such as OBJ_BULK_LIMIT/4) via bulk:

{noformat}
ERROR: daos_io_server:0 *** Process 129953 received signal 11 ***
ERROR: daos_io_server:0 Associated errno: Success (0)
Failing for address: 0x18bf740
ERROR: daos_io_server:0 /lib64/libpthread.so.0(+0xf5e0)[0x7fa8614f55e0]
ERROR: daos_io_server:0 /lib64/libc.so.6(+0x142de5)[0x7fa8609bfde5]
/home/fyong/Work/DAOS/D4/daos/install/lib64/../prereq/dev/mercury/lib/libmercury.so.2(HG_Bulk_serialize+0x204)[0x7fa85fbf5494]
/home/fyong/Work/DAOS/D4/daos/install/lib64/../prereq/dev/mercury/lib/libmercury.so.2(hg_proc_hg_bulk_t+0x110)[0x7fa85fbfff60]
ERROR: daos_io_server:0 /home/fyong/Work/DAOS/D4/daos/install/lib64/libcart.so.4(crt_proc_crt_bulk_t+0x9)[0x7fa8623733f9]
/home/fyong/Work/DAOS/D4/daos/install/lib64/libdaos.so.0(+0xa893e)[0x7fa861dee93e]
/home/fyong/Work/DAOS/D4/daos/install/lib64/libdaos.so.0(+0xaa438)[0x7fa861df0438]
ERROR: daos_io_server:0 /home/fyong/Work/DAOS/D4/daos/install/lib64/libdaos.so.0(+0xac1c4)[0x7fa861df21c4]
/home/fyong/Work/DAOS/D4/daos/install/lib64/libcart.so.4(crt_proc_in_common+0x18d)[0x7fa862374e8d]
/home/fyong/Work/DAOS/D4/daos/install/lib64/../prereq/dev/mercury/lib/libmercury.so.2(+0x538d)[0x7fa85fbef38d]
/home/fyong/Work/DAOS/D4/daos/install/lib64/../prereq/dev/mercury/lib/libmercury.so.2(HG_Forward+0x5f)[0x7fa85fbf1a5f]
/home/fyong/Work/DAOS/D4/daos/install/lib64/libcart.so.4(crt_hg_req_send+0x159)[0x7fa86236cf59]
/home/fyong/Work/DAOS/D4/daos/install/lib64/libcart.so.4(crt_req_send_internal+0x178)[0x7fa8623ac5d8]
/home/fyong/Work/DAOS/D4/daos/install/lib64/libcart.so.4(+0x91484)[0x7fa8623ae484]
ERROR: daos_io_server:0 /home/fyong/Work/DAOS/D4/daos/install/lib64/libcart.so.4(crt_req_send_internal+0x7e3)[0x7fa8623acc43]
/home/fyong/Work/DAOS/D4/daos/install/lib64/libcart.so.4(+0x933cc)[0x7fa8623b03cc]
/home/fyong/Work/DAOS/D4/daos/install/lib64/libcart.so.4(+0x4a310)[0x7fa862367310]
/home/fyong/Work/DAOS/D4/daos/install/lib64/../prereq/dev/mercury/lib/libmercury.so.2(+0x4671)[0x7fa85fbee671]
/home/fyong/Work/DAOS/D4/daos/install/lib64/../prereq/dev/mercury/lib/libmercury.so.2(+0xfa8b)[0x7fa85fbf9a8b]
/home/fyong/Work/DAOS/D4/daos/install/lib64/../prereq/dev/mercury/lib/libmercury.so.2(HG_Core_trigger+0xb)[0x7fa85fbff38b]
/home/fyong/Work/DAOS/D4/daos/install/lib64/../prereq/dev/mercury/lib/libmercury.so.2(HG_Trigger+0x12)[0x7fa85fbf1dc2]
/home/fyong/Work/DAOS/D4/daos/install/lib64/libcart.so.4(crt_hg_progress+0x72)[0x7fa86236ffd2]
/home/fyong/Work/DAOS/D4/daos/install/lib64/libcart.so.4(crt_progress+0x25)[0x7fa862333675]
/home/fyong/Work/DAOS/D4/daos/install/bin/daos_io_server[0x41cc27]
/home/fyong/Work/DAOS/D4/daos/install/bin/../prereq/dev/argobots/lib/libabt.so.0(+0x131bb)[0x7fa8612d71bb]
/home/fyong/Work/DAOS/D4/daos/install/bin/../prereq/dev/argobots/lib/libabt.so.0(+0x13891)[0x7fa8612d7891]
instance 0 exited: /home/fyong/Work/DAOS/D4/daos/install/bin/daos_io_server (instance 0) exited: signal: segmentation fault
{noformat}

I can adjust client side logic to avoid using small sized data via bulk transfer. But it is not the finial solution. The root reason is inside mercury for small sized data bulk transfer.

**Comment by frostedcmos** [comment]
Thanks, I've gathered information from Nasf-Fan and submitted issue with mercury team

**Review by jolivier23 — CHANGES_REQUESTED** [review]
we cut the tag but there is a conflict on the spec

**Comment by Nasf-Fan** [comment]
The daos side solution to avoid using small sized data bulk transfer is here:
https://github.com/daos-stack/daos/pull/4012

**Pull request closed without merging**